Add new feature to find file(recreate)

koreamic opened the pull request:
Using the fuzzy lib, develop "file finder" feature.
(http://feedback.gitlab.com/forums/176466-general/suggestions/4987909-add-file-finder-fuzzy-input-in-files-tab-to-ju)

Issue(#9854)
PR(#9855)
- apply ajax request
- fuzzy filter(https://github.com/jeancroy/fuzzaldrin-plus)
- short cut(when "t" was hitted, go to 'file find' page and 'esc' is to go back)
- feature screenshot
  ![file_finder](https://cloud.githubusercontent.com/assets/14069553/11911666/93930a72-a660-11e5-8693-5a3ccf33eb85.gif)

(ps. cause by my mistake, recreate PR(#9855), sorry!!)

stanhu commented:
Great work! I assume since this depends on [this gitlab_git MR](https://gitlab.com/gitlab-org/gitlab_git/merge_requests/60), this PR has to wait until a new version is release?

Can you add tests for this?

koreamic committed:
Add new feature to find file

Using the fuzzy filter, develop "file finder" feature.
- feedback(http://feedback.gitlab.com/forums/176466-general/suggestions/4987909-add-file-finder-fuzzy-input-in-files-tab-to-ju )
- fuzzy filter(https://github.com/jeancroy/fuzzaldrin-plus)
- shortcuts(when "t" was hitted at tree view, go to 'file find' page and 'esc' is to go back)
- depends on gitlab_git 7.2.22

stanhu commented on app/controllers/projects/find_file_controller.rb:
```diff
@@ -0,0 +1,26 @@
+  include ExtractsPath
+  include ActionView::Helpers::SanitizeHelper
+  include TreeHelper
+  before_action :assign_ref_vars
```
Do we also need to add:

```ruby
  before_action :require_non_empty_project
  before_action :authorize_download_code!
```

stanhu commented:
Does this PR also need an upgrade of gitlab_git to 7.2.22?

stanhu commented:
Also, this feature could use a spinach test to test the JavaScript functionality.

koreamic commented:
@stanhu
Thank you for your concern and help.

Yes. This PR  needs an upgrade of gitlab_git to 7.2.22.

and okay I will check to use a spinach test.

If you need to additional works. please comment.

Thank you. have a nice day.

Razer6 commented on app/assets/javascripts/shortcuts_navigation.coffee:
```diff
@@ -14,6 +14,7 @@ class @ShortcutsNavigation extends Shortcuts
     Mousetrap.bind('g m', -> ShortcutsNavigation.findAndFollowLink('.shortcuts-merge_requests'))
     Mousetrap.bind('g w', -> ShortcutsNavigation.findAndFollowLink('.shortcuts-wiki'))
     Mousetrap.bind('g s', -> ShortcutsNavigation.findAndFollowLink('.shortcuts-snippets'))
+    Mousetrap.bind('t', -> ShortcutsNavigation.findAndFollowLink('.shortcuts-find-file'))
```
Shouldn't this only be active to on the tree page? Can you also update https://github.com/gitlabhq/gitlabhq/blob/master/app/views/help/_shortcuts.html.haml and the screenshot in https://github.com/gitlabhq/gitlabhq/blob/master/doc/workflow/shortcuts.md

koreamic replied:
Okay, @Razer6 Thank you for your suggestions.
I updated ```_shortcuts.html.haml``` and screenshot.
and I added shortcuts using up and down array key in find file view to select file more conveniently.

The reason that add ```t``` shortcut in ```app/assets/javascripts/shortcuts_navigation.coffee``` is to use find file view other page(eg. merge request/commit/branch page) 
If necessary.

but If not necessary, I just define it on the tree view.

what do you think about that?

DouweM replied:
I think it doesn't make sense on the issues page, but it does on any file pages (tree, blob, blame) as well as the project root.

koreamic commented:
@stanhu 
I added spinach test case and shortcuts that up down arrow to select item easly.
and modified some bugs(main menu bar out focusing)

please check and review sources

thank you.

yms9654 commented:
:+1: This is a great feature!
I think many users like this.

randx commented on app/assets/javascripts/lib/fuzzaldrin-plus.js:
```diff
@@ -0,0 +1,1158 @@
+(function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;fo ...
```
Please move this file to `/vendor/assets/javascripts`

randx commented:
Please provide a screenshots for this feature

stanhu commented on app/assets/javascripts/project_find_file.js.coffee:
```diff
@@ -0,0 +1,110 @@
+
+  # make tbody row html
+  makeHtml: (filePath, blobItemUrl) ->
+    return  @row.replace("%blobItemUrl", blobItemUrl).replace("%filePath", filePath)
```
What happens if the file path has the text `%blobItemUrl`? Can we just construct the entire text without a `replace`?

DouweM replied:
As mentioned above, please generate the row DOM using jQuery.

stanhu commented on app/assets/javascripts/project_find_file.js.coffee:
```diff
@@ -0,0 +1,110 @@
+  # go back page
+  goBack: (event) ->
+    if event.keyCode == 27
+      event.preventdefault
```
Isn't this `event.preventDefault()`?

stanhu commented on app/assets/javascripts/project_find_file.js.coffee:
```diff
@@ -0,0 +1,110 @@
+    return false;
+
+  # select row and focus
+  selecteRow: (rows, seletedRow, which) ->
```
Is this a misspelling: `selectedRow`?

stanhu commented on app/assets/javascripts/project_find_file.js.coffee:
```diff
@@ -0,0 +1,110 @@
+    $(document).on "keyup", @element, (event) =>
+      keyCode = event.keyCode if event
+      # up down right left arrow(37~40) esc(27) caps(20) ctrl(17) alt(18) shift(16) enter(13) cmd(91)
+      if (keyCode < 37 || keyCode > 40) && keyCode != 20 && keyCode != 17 && keyCode != 18 && keyCode != 16 && keyCode != 13 && keyCode != 91
```
Should we define these key codes as constants (e.g. https://github.com/cybertoothca/keyevent/blob/master/src/keyevent.js)?

DouweM commented on app/assets/javascripts/project_find_file.js.coffee:
```diff
@@ -0,0 +1,109 @@
+  constructor: (@options)->
+    @filePaths = {}
+    @element = ".file-finder-input"
+    @row = "<tr class='tree-item'><td class='tree-item-file-name'><i class='fa fa-file-text-o fa-fw'></i><span class='str-truncated'><a href=%blobItemUrl>%filePath</a></span></td></tr>"
```
Please build this structure using jQuery. Doing it as a string opens us up to XSS.

DouweM commented on app/assets/javascripts/project_find_file.js.coffee:
```diff
@@ -0,0 +1,109 @@
+  # init key event
+  initKeyEvent: ->
+    # unbind and bind keyup/keydown event at text input box
+    $(document).off "keyup"
```
I don't think we want to get rid of all keyup events. Can you namespace these to be specific to `ProjectFindFile`? 

From https://api.jquery.com/bind/:

> If the eventType string contains a period (.) character, then the event is namespaced. The period character separates the event from its namespace. For example, in the call .bind( "click.name", handler ), the string click is the event type, and the string name is the namespace. Namespacing allows us to unbind or trigger some events of a type without affecting others. See the discussion of .unbind() for more information.

DouweM commented on app/assets/javascripts/project_find_file.js.coffee:
```diff
@@ -0,0 +1,109 @@
+    # $(document).on "keyup", @goBack
+    $(document).on "keyup", (event) =>
+      if event.keyCode == 27
+        event.preventdefault
```
`event.preventDefault()`

DouweM commented on app/assets/javascripts/project_find_file.js.coffee:
```diff
@@ -0,0 +1,109 @@
+      @renderList result, searchTxt
+
+  # select row and focus
+  selecteRow: (rows, seletedRow, which) ->
```
- `selectRow`
- `selectedRow`

DouweM commented on app/assets/javascripts/project_find_file.js.coffee:
```diff
@@ -0,0 +1,109 @@
+
+      # 38 up 40 down
+      if keyCode == 38 || keyCode == 40
+          @selecteRow $("tr.tree-item"), $('.tree-item.selected'), keyCode
```
Inconsistent indentation

DouweM commented on app/assets/javascripts/project_find_file.js.coffee:
```diff
@@ -0,0 +1,109 @@
+
+  # find file
+  findFile: (searchTxt) ->
+      result = if searchTxt then fuzzaldrinPlus.filter(@filePaths, searchTxt) else @filePaths
```
Inconsistent indentation

DouweM commented on app/assets/javascripts/project_find_file.js.coffee:
```diff
@@ -0,0 +1,109 @@
+    $(document).on "keyup", @element, (event) =>
+      keyCode = event.keyCode if event
+      # up down right left arrow(37~40) esc(27) caps(20) ctrl(17) alt(18) shift(16) enter(13) cmd(91)
+      if (keyCode < 37 || keyCode > 40) && keyCode != 20 && keyCode != 17 && keyCode != 18 && keyCode != 16 && keyCode != 13 && keyCode != 91
```
Why are we checking all these keycodes?

Razer6 replied:
We already use moustrap (https://github.com/ccampbell/mousetrap), is this keycode checking necessary with that?

DouweM commented on app/assets/javascripts/project_find_file.js.coffee:
```diff
@@ -0,0 +1,109 @@
+      # 38 up 40 down
+      if keyCode == 38 || keyCode == 40
+          @selecteRow $("tr.tree-item"), $('.tree-item.selected'), keyCode
+          return false
```
No need to return since there are no statements after this if/else.

DouweM commented on app/assets/javascripts/project_find_file.js.coffee:
```diff
@@ -0,0 +1,109 @@
+      method: "get"
+      dataType: "json"
+      success: $.proxy((data) ->
+        $(".loading").hide()
```
We need to make this more specific in case we get other `.loading`s in the future.

DouweM commented on app/assets/javascripts/project_find_file.js.coffee:
```diff
@@ -0,0 +1,109 @@
+      url: url
+      method: "get"
+      dataType: "json"
+      success: $.proxy((data) ->
```
Why do we need $.proxy?

DouweM commented on app/assets/javascripts/project_find_file.js.coffee:
```diff
@@ -0,0 +1,109 @@
+        $(".loading").hide()
+        @filePaths = data
+        @findFile($(@element).val())
+        @selecteRow $("tr.tree-item")
```
We need to make all of these more specific :) Can we pass an element to `new ProjectFindFile`, and then use `@element.find('tr.tree-item')` etc?

DouweM commented on app/assets/javascripts/project_find_file.js.coffee:
```diff
@@ -0,0 +1,109 @@
+  renderList: (filePaths, searchTxt) ->
+    $(".tree-table > tbody").empty()
+    for filePath, i in filePaths
+      break if i > 20
```
This means we break when `i == 21`, which means we had 0..20 before this: 21 renders :) We need only 20.

DouweM commented on app/assets/javascripts/project_find_file.js.coffee:
```diff
@@ -0,0 +1,109 @@
+  selecteRow: (rows, seletedRow, which) ->
+    if rows && rows.length > 0
+      if seletedRow && seletedRow.length > 0
+        next = null
```
Why is this necessary?

DouweM commented on app/assets/javascripts/project_find_file.js.coffee:
```diff
@@ -0,0 +1,109 @@
+      $(".tree-table > tbody").append(html)
+
+  # highlight text(awefwbwgtc -> <b>a</b>wefw<b>b</b>wgt<b>c</b> )
+  highlightText = (txt, charArr) ->
```
This is the highlighting code that the Atom editor uses: https://github.com/atom/fuzzy-finder/blob/master/lib%2Ffuzzy-finder-view.coffee#L70-L88 Maybe we can use the same logic :)

DouweM commented on app/controllers/projects/find_file_controller.rb:
```diff
@@ -0,0 +1,30 @@
+
+    @url = namespace_project_files_path(@project.namespace, @project, @ref, @options.merge(format: :json))
+    @treeUrl = namespace_project_tree_path(@project.namespace, @project, @ref)
+    @blobUrlTemplate = namespace_project_blob_path(project.namespace, project, tree_join(@id || @commit.id))
```
We don't need all of these in the controller, we can just use the helpers in the view.

DouweM commented on app/controllers/projects/find_file_controller.rb:
```diff
@@ -0,0 +1,30 @@
+  end
+
+  def list
+    filePaths = @repo.ls_files(@ref)
```
Ruby wants `snake_case`, not `camelCase` :)

DouweM commented on app/views/projects/_find_file_link.html.haml:
```diff
@@ -0,0 +1,6 @@
+    %span.btn-group{class: 'btn-group pull-right hidden-xs hidden-sm'}
+      = link_to namespace_project_find_file_path(project_namespace, project, ref), class: 'btn btn-success shortcuts-find-file', rel: 'nofollow' do
+        = icon('search')
+        %span FIND FILE
```
This can be `Find File`, it will automatically be rendered all caps :)

DouweM commented on app/views/projects/_find_file_link.html.haml:
```diff
@@ -0,0 +1,6 @@
+- if project && project_namespace && ref
+  .find-file-link-holder
+    %span.btn-group{class: 'btn-group pull-right hidden-xs hidden-sm'}
+      = link_to namespace_project_find_file_path(project_namespace, project, ref), class: 'btn btn-success shortcuts-find-file', rel: 'nofollow' do
```
I don't think we need `btn-success` here.

DouweM commented on app/views/projects/_find_file_link.html.haml:
```diff
@@ -0,0 +1,6 @@
+- if project && project_namespace && ref
```
We can just use `@project` and `@ref` directly here, and we don't need to verify that they exist since they always do on the tree page :)

DouweM commented:
@koreamic What happened to Network shortcuts in the shortcuts screenshot? :)

koreamic commented:
Hi. Happy new year.

@DouweM , @stanhu , @randx , @Razer6 

I updated sources for this feature. and I accepted most of your advice.

- updated shortcuts screenshot
- used mousetrap for shortcuts(up/down/esc/enter, avoid to bind and unbind at ```$(document)``` event)
- removed most of key codes
- highlighting path logic
- wrapper element
- other small modifications(indentations, button class, unnecessary codes ... )

thank you.

DouweM commented on app/assets/javascripts/project_find_file.js.coffee:
```diff
@@ -0,0 +1,121 @@
+  # find file
+  findFile: ->
+    searchTxt = @inputElement.val()
+    result = if searchTxt then fuzzaldrinPlus.filter(@filePaths, searchTxt) else @filePaths
```
When `searchTxt` equals `""`, it will still be true and we enter fuzzAldrin. Please check if `searchTxt.length > 0`

DouweM commented on app/assets/javascripts/project_find_file.js.coffee:
```diff
@@ -0,0 +1,121 @@
+
+  # find file
+  findFile: ->
+    searchTxt = @inputElement.val()
```
`searchTxt` is not much shorter than `searchText`, please use full variable names :)

DouweM commented on app/assets/javascripts/project_find_file.js.coffee:
```diff
@@ -0,0 +1,121 @@
+        @filePaths = data
+        @findFile()
+        @element.find('#files-slider tr.tree-item').eq(0).addClass("selected").focus()
+      ).bind(this)
```
You should be able to use CoffeeScript's `=>` instead of `bind(this)`

DouweM commented on app/assets/javascripts/project_find_file.js.coffee:
```diff
@@ -0,0 +1,121 @@
+        @element.find(".loading").hide()
+        @filePaths = data
+        @findFile()
+        @element.find('#files-slider tr.tree-item').eq(0).addClass("selected").focus()
```
Can we not use ids on elements, but only classes instead?

DouweM commented on app/assets/javascripts/project_find_file.js.coffee:
```diff
@@ -0,0 +1,121 @@
+
+  # render result
+  renderList: (filePaths, searchTxt) ->
+    $(".tree-table > tbody").empty()
```
Please find this within `@element` or something.

DouweM commented on app/assets/javascripts/project_find_file.js.coffee:
```diff
@@ -0,0 +1,121 @@
+    $(".tree-table > tbody").empty()
+
+    for filePath, i in filePaths
+      break if i > 19
```
So if `i == 20`? :) That may be a little clearer.

DouweM commented on app/assets/javascripts/project_find_file.js.coffee:
```diff
@@ -0,0 +1,121 @@
+      break if i > 19
+      highlightFilePath = filePath
+
+      if searchTxt
```
`searchTxt.length > 0`

DouweM commented on app/assets/javascripts/project_find_file.js.coffee:
```diff
@@ -0,0 +1,121 @@
+
+      blobItemUrl = "#{@options.blobUrlTemplate}/#{filePath}"
+      html = @makeHtml highlightFilePath, blobItemUrl
+      $(".tree-table > tbody").append(html)
```
Please use `@element.find` or similar.

DouweM commented on app/assets/javascripts/project_find_file.js.coffee:
```diff
@@ -0,0 +1,121 @@
+
+  # make tbody row html
+  makeHtml: (filePath, blobItemUrl) ->
+    return $("<tr class='tree-item'><td class='tree-item-file-name'><i class='fa fa-file-text-o fa-fw'></i><span class='str-truncated'><a href=#{blobItemUrl}>#{filePath}</a></span></td></tr>")
```
To prevent XSS when `blobItemUrl` or `filePath` contain something that looks like HTML/JS, Build the `span` element like you do now, and then create the link like `$('<a>').attr('href', blobItemUrl).text(filePath)` and add it to the span using jQuery.

DouweM replied:
Also, no need to explicitly return in CoffeeScript :)

DouweM commented on app/assets/javascripts/project_find_file.js.coffee:
```diff
@@ -0,0 +1,121 @@
+    location.href = @options.treeUrl
+
+  goToBlob: =>
+    path = $('.tree-item.selected .tree-item-file-name a').attr('href')
```
`@element.find` please!

DouweM commented on app/assets/javascripts/project_find_file.js.coffee:
```diff
@@ -0,0 +1,121 @@
+
+  goToBlob: =>
+    path = $('.tree-item.selected .tree-item-file-name a').attr('href')
+    if path then location.href = path
```
`location.href = path if path` is more CoffeeScript-like.

DouweM commented on app/assets/javascripts/shortcuts_find_file.js.coffee:
```diff
@@ -0,0 +1,15 @@
+    Mousetrap.stopCallback = (event, element, combo) =>
+      if combo == 'up' or combo == 'down' or combo == 'esc' or combo == 'enter'
+        return element != @projectFindFile.inputElement[0]
+      return _oldStopCallback(event, element, combo)
```
Why do we need to override `Mousetrap.stopCallback`?

koreamic replied:
@DouweM 
When focus in textbox, Mousetrap's shortcuts is not working.
I think that we want to fire shortcuts action at that time in find file page.
so I overrided `Mousetrap.stopCallback`
and added `event.preventDefault` for avoiding to move cusor in textbox.
(when up/down key pressed, cusor is moved to home/end)

DouweM commented on app/views/projects/tree/show.html.haml:
```diff
@@ -3,12 +3,12 @@
 - if can? current_user, :download_code, @project
   .tree-download-holder
     = render 'projects/repositories/download_archive', ref: @ref, btn_class: 'btn-group pull-right hidden-xs hidden-sm', split_button: true
+= render 'projects/find_file_link'
```
We need a little bit of distance between the "Find File" and "Download zip" buttons. This works:

- Change `_find_file_link.html` to:

```haml
= link_to namespace_project_find_file_path(@project.namespace, @project, @ref), class: 'btn btn-grouped shortcuts-find-file', rel: 'nofollow' do
  = icon('search')
  %span Find File
```

- Change in `tree/show.html.haml` from:

```haml
- if can? current_user, :download_code, @project
  .tree-download-holder
    = render 'projects/repositories/download_archive', ref: @ref, btn_class: 'btn-group pull-right hidden-xs hidden-sm', split_button: true
= render 'projects/find_file_link'
```
to:
```haml
.pull-right
  = render 'projects/find_file_link'
  - if can? current_user, :download_code, @project
    = render 'projects/repositories/download_archive', ref: @ref, btn_class: 'hidden-xs hidden-sm btn-grouped', split_button: true
```

DouweM commented:
@koreamic I may seem very critical (and I am), but I tried it out locally and it works really, really well. Great job. I can't wait to have this in the next release :) I'm sure we can get it ready.

koreamic commented:
@DouweM 
I have to thank you for your review and concern all the time.
I have learned many things.
I modified sources and commited as your review.
but If there are things to be modified, leave comments.

DouweM commented:
@koreamic You're amazing. I've got nothing, I'm merging this.

JobV commented:
This is awesome. Thanks @koreamic

DouweM commented:
@koreamic I notice you're not credited in the changelog! I added `(koreamic)` in https://gitlab.com/gitlab-org/gitlab-ce/merge_requests/2327, let me know if you want that to have your full name instead.

Razer6 commented:
:heart: t

stanhu commented:
Great work, @koreamic!

koreamic commented:
@DouweM @stanhu @Razer6 @randx @JobV @yms9654 @noveler17
Thanks a lot for your help & patience.
@DouweM `koreamic` is enough. thanks

randx commented:
@koreamic can you please provide us with your full name? We prefer not to use usernames since it can be different between websites and can produce unnecessary confusion.  cc @DouweM

randx commented:
@koreamic and thank you for getting this feature done! :sparkles:

koreamic commented:
@randx @DouweM 
Okay. My full name is `Kyungchul Shin`
Thank you.

randx commented:
@koreamic thank you! 

@DouweM I will update CHANGELOG

randx commented:
wow @stanhu already updated CHANGELOG :smile: :rocket:

tsigo commented:
@koreamic Curious, why did you use fuzzaldrin-plus instead of fuzzaldrin?

DouweM commented:
@tsigo That was at my suggestion actually. I had much better results in Atom's fuzzy finder with the "Use Alternate Scoring" option enabled, which switches from fuzzaldrin to fuzzaldrin-plus. It's now the default in Atom's fuzzy finder as well: https://github.com/atom/fuzzy-finder/commit/c004c7ec19c631fbc3d01c4fd4a24fb95f9b601f

tsigo commented:
Ok. Do we know which version it was? I'm trying to replace the minified vendor JS with non-minified versions (specifically so we know version info, among other things).

DouweM commented:
@tsigo Ah, good call. It doesn't really matter, the API hasn't changed, so you can just get the latest one.

johnleeleelee commented:
Ok. Do we know which version it was? I'm trying to replace the minified vendor JS with non-minified versions (specifically so we know version info, among other things).
[https://www.jqueenwatchwinders.co.uk/best-jqueen-watch-winders-for-rolex
](url)
